### ISSUE TITLE
LaTeX fails on description of С function with arguments

### ISSUE THREAD

**Issue by bunder** [sphinx-doc/sphinx#3019]
Sphinx 1.4.8. Compilation of LaTeX fails when using `c:function` for function with arguments.
In ReST:

```
.. c:function:: int foo(int a)

   My foontion.
```

In LaTeX file:

```
\index{foo (C function)}

\begin{fulllineitems}
\phantomsection\label{abc:c.foo}\pysiglinewithargsret{int \sphinxbfcode{foo}}{int\emph{ a}}{}
My foonction.

\end{fulllineitems}
```

Upon compilation:

```
! Missing } inserted.
<inserted text> 
                }
l.536 ...{int \sphinxbfcode{foo}}{int\emph{ a}}{}
?
```

**Comment by jfbu** [comment]
I can not reproduce the error with Sphinx 1.4.8 or the master branch. In the former case the LaTeX file does contain

``` latex
\begin{document}

\maketitle
\tableofcontents
\phantomsection\label{index::doc}

\index{foo (C function)}

\begin{fulllineitems}
\phantomsection\label{index:c.foo}\pysiglinewithargsret{int \sphinxbfcode{foo}}{int\emph{ a}}{}
My foontion.

\end{fulllineitems}




\renewcommand{\indexname}{Index}
\printindex
\end{document}
```

and there is no error on LaTeX compilation. The master branch has slightly different LaTeX code but no error either.

**Comment by jfbu** [comment]
can you reproduce the error with  a minimal project ?

**Comment by jfbu** [comment]
@bunder do you still have this issue or can we close the ticket ?

**Comment by bunder** [comment]
Sorry for being silent for a while, I'll try to reproduce the problem on a minimal project.

**Comment by tk0miya** [comment]
I can't reproduce the problem too.

**Comment by bunder** [comment]
Ok, I can not reproduce the problem with minimal project. However, I've found the problem origin in my environment.
We are using `xelatex` with following preamble (this snippet [can be found all over internet](https://www.google.com/search?q=setTransitionsForLatin))

```
\usepackage{ucharclasses}

\setmainlanguage{russian}
\setdefaultlanguage{russian}
\setotherlanguages{english}

\setTransitionsForLatin{\begingroup\hyphenrules{english}}{\endgroup}
```

to produce PDF. With these settings code `{int\emph{ a}}` (generated by `c:function`, space is a non-breaking space) rises an error, while the code `{int \emph{a}}` (generated by `cpp:function`, space is an ordinary space) does not.
I'll fix my environment, however I think it's worth fixing the code generated by `c:function`, because
1. It is better when `c:function` and `cpp:funtion` generate nearly the same code.
2. LaTeX has special character `~` for non-breaking space, which is far more portable.
3. IMHO, including space under the _emphasis_ `{int\emph{ a}}` is rather strange.

**Comment by jfbu** [comment]
Ok, thanks for the added explanations. I had not paid attention to the non-breaking space. 

It originates in file https://github.com/sphinx-doc/sphinx/blob/master/sphinx/domains/c.py#L165, and was introduced at 0aa953848.

I agree with your 2 and 3, and I trust 1 also makes sense, but not being much a practitioner I can't tell. 

 @birkenfeld what do you think ?

**Comment by jfbu** [comment]
I forgot to comment that yes `~` is standard in LaTeX, but as you use `xetex`,  using a non-breaking space should work too.

And more importantly, the nodes here at examined at earlier time than writer output. Putting a `~` is only good for LaTeX and it is probably not an option for [c.py](https://github.com/sphinx-doc/sphinx/blob/master/sphinx/domains/c.py).

**Comment by jfbu** [comment]
It appears that there is either some bug in `ucharclasses` package or in the way it is used. The code:

``` latex
\documentclass[russian]{article}
\usepackage{fontspec}
\setmainfont{DejaVu Serif}
\setsansfont{DejaVu Sans}
\setmonofont{DejaVu Sans Mono}

\usepackage[english,russian]{babel}
% \setmainlanguage{russian}
% \setdefaultlanguage{russian}
% \setotherlanguages{english}

\usepackage{ucharclasses}
\setTransitionsForLatin{\begingroup}{\endgroup}% BAD
%%\setTransitionsForLatin{\quad}{\quad}% OK

\begin{document}

\selectlanguage{russian}

функция C\begin{otherlanguage}{english}
  abcd\end{otherlanguage} a% non-breaking space used before the a


\end{document}
% Local Variables:
% TeX-engine: xetex
% End:
```

(I use Babel above because I was doing tests related to `\hyphenrules`, but same with polyglossia)

fails to compile on both TL2015 and TL2016. But if I comment the BAD line
and un-comment the OK line it compiles.

Are you sure this `\setTransitionsForLatin{\begingroup}{\endgroup}` is licit ? Perhaps this is a bug of `ucharclasses`, you should ask on tex.stackexchange.com.

I am sorry I can't help much here as I am not at all familiar with polyglossia+ucharclasses.

**Comment by jfbu** [comment]
actually now that I have removed the `\emph` the error message is clearer

```
! Extra \endgroup.
<XeTeXinterchartoks> \endgroup 

l.21   abcd\end{otherlanguage} a

? 
```

it does look like opening and closing groups in `setTransitionsForLatin` is either illegal or it is legal and `ucharclasses` has a bug.

**Comment by tk0miya** [comment]
For pdftex, non-breaking space was supported at 9cfdf9b13eaf5cc70e1f83f11e8d9155b2677671.
It was replaced to `\nobreakspace` macro using `\DeclareUnicodeCharacter`.

Is it needed to move the replacement to latex writer?

**Comment by jfbu** [comment]
@tk0miya no I think replacement should not be moved to latex writer, because currently since 83da5df, the `\DeclareUnicodeCharacter` is correctly wrapped into a conditional to not interfere with `xelatex` and `lualatex` which handle natively the Unicode no-break space, _assuming fontspec+OpenType fonts are used_ . 

<strike>There are I think some issues with other Unicode space characters, such as "narrow no-break space" which need extra steps, for example I have in a project of mine using `lualatex`:

```
    'utf8extra':
'''\\catcode`^^^^202f\\active
\\def^^^^202f{\\nobreak\\FBthinspace}''',
```

to define how the narrow nobreak space acts in latex (`\FBthinspace` is a thin space configured by French language).</strike>

<strike>But</strike> regarding 00a0, there does not seem to be a real issue. The problem in this thread is one of use of `ucharclasses` package for xetex users.

[This project](https://github.com/wilx/project-uspace) seems to be about making all Unicode space characters work in all three engines.

**Comment by jfbu** [comment]
a precision: my conf.py extract to define narrow no-break space was not needed for intrinsic reasons but only to redefine what it does. Without any definition, assuming you use LuaLaTeX or XeLaTeX with an OpenType font which has the character, it does work out of the box.

**Comment by jfbu** [comment]
Thus, the issue here is not one of Sphinx per se as it produces correct LaTeX code.

There remains the matter of making `c:function::` and `cpp:function::`  generate more similar code. I am afraid however that examination of this may lead to more Unicode nobreak space characters being inserted ;-)

**Comment by jfbu** [comment]
@bunder Can you try to use the `'utf8extra'` key  in your conf.py like this ? 

```
latex_engine= 'xelatex'

latex_elements = {
     # The paper size ('letterpaper' or 'a4paper').
     #
     # 'papersize': 'letterpaper',

     # The font size ('10pt', '11pt' or '12pt').
     #
     # 'pointsize': '10pt',

     # Additional stuff for the LaTeX preamble.
     #
     # 'preamble': '',

'babel': '\\usepackage{polyglossia}',

'utf8extra':
'''\\catcode`^^^^00a0\\active\\def^^^^00a0{\\nobreakspace}''',

'fontenc': '''\
\\usepackage{fontspec}
\\setmainfont{DejaVu Serif}
\\setsansfont{DejaVu Sans}
\\setmonofont{DejaVu Sans Mono}
''',

     # Latex figure (float) alignment
     #
     # 'figure_align': 'htbp',
'preamble' : r'''\usepackage{ucharclasses}

\setmainlanguage{russian}
\setdefaultlanguage{russian}
\setotherlanguages{english}

\setTransitionsForLatin{\begingroup\hyphenrules{english}}{\endgroup}
''',
}
```

This seems to go around the problem created by `ucharclasses` and may hep you compile your documents to PDF with minimal change.

**addendum:** if your environment loads explicitly old package `xunicode` (which is not recommended nowadays) at some point, an extra step is needed to circumvent it. You should add `\AtBeginDocument{\DeclareTextCommandDefault{\nobreakspace}{\leavevmode\nobreak\ }}` to your `'preamble'` key (or put the `\Declare...` anywhere after loading of `xunicode`).

**Comment by jakobandersen** [comment]
> There remains the matter of making c:function:: and cpp:function:: generate more similar code. I am afraid however that examination of this may lead to more Unicode nobreak space characters being inserted ;-)

I'm (slowly) working on a new version of the C domain based on the C++ domain, so equivalent declarations should then generate equivalent code.

**Comment by jfbu** [comment]
@tk0miya in this thread I have said that Unicode nobreak space works out of the box in xetex/luatex with OpenType fonts. But this leaves possibly erroneous impression:
- yes indeed, a nobreak space character does act as a space which prevents breaks,
- but it is not like LaTeX's standard `~` because the latter has some stretch/shrink which the Unicode nobreak space character does **not** have.

It is indeed possible to make a Unicode nobreak space act like the `~` using 

```
\usepackage{newunicodechar}
\newunicodechar{ }{\nobreakspace}% the space is a nobreak space
```

but if one looks at what this does, it defines a long sequence of intermediate things hence a more efficient way is

```
\usepackage{newunicodechar}
\newunicodechar{ }{\leavevmode\nobreak\ }
```

but this needs the relatively recent package `newunicodechar` (2012).

We can code directly this:

```
\catcode`^^^^00a0\active\protected\def^^^^00a0{\leavevmode\nobreak\ }
```

If we want that by default Unicode nobreak space character is like LaTeX's `~`, then I think this should be incorporated into LaTeX template when the engine is `lualatex` or `xelatex`.

**Comment by jfbu** [comment]
@tk0miya unbeknown to me the `C` in the issue title was a Unicode letter, it made its way to CHANGES at 5caa6c3, breaking build of pdf doc, and I fixed it later at 04e1e7d.
